site/content/blog/scraping-with-chatgpt/index.md

```diff
@@ -0,0 +1,220 @@
+LangChain](https://github.com/hwchase17/langchainjs),
+too. OpenAI libraries are 
+coming 
+out in just 
+about every
+popular language now, or you could easily write your own client. Nothing in here is very complicated or qualifies as
+real "machine learning". To be honest, sometimes it doesn't even feel like programming. It's that easy.
+
+
+##### Looping
+Because of the token
+limit of 4096 (which tends to be around 16000 chars), pages are often too long to pass in all at once. Note that in 
+the prompt, we can pass in the end of a previous section, and a previous summary. That's so that we can loop ChatGPT 
+over chunks of the document, and then in the end we just concat them all together. 
+
+There is a choice between calling ChatGPT in parallel (faster) or in sequence (slower), and in this case we chose to 
+go in sequence, map-reduce style. This is so that we can have ChatGPT stitch the chunks together in a grammatically 
+correct way. Otherwise, the chunk boundaries might get a bit garbled. In other words, we are chunking the scraped 
+web page contents into ChatGPT, with a little bit of an overlapping window where the last result feeds into the next 
+query.
+This method also helps refine the 
+summary we asked ChatGPT to give us as it reads down 
+the page, adding new information without losing context from higher up the page. 
+
+##### How to calculate your token budget
+Going over the 4096 token limit in one request will break the entire flow.
+The token limit is a bit like playing blackjack.
+Ideally we will nearly fill it up, but it's better to be under than over.
+
+
+ Note that we have to 
+leave space for the prompt, the passed in page contents, the snippets from previous loops, **and ChatGPT's 
+response**. We know that in this case the output will be roughly the same size or smaller than the input size, so if 
+we split our roughly 4000 token budget in half, that leaves us 2000 tokens for input. Since our prompt takes us 
+about 500 tokens, let's just use a size of 1500 as our splitter. You can use
+[OpenAI's tokenizer](https://platform.openai.com/tokenizer) to see how many tokens your prompt uses. The 
+ visualization is nice as well and gives you a feel for how tokens work.
+
+To split the documents in an intelligent way, we are using Langchain's `TokenTextSplitter` to give us an array of 
+chunks of the proper size. 
+
+```python
+from langchain.text_splitter import TokenTextSplitter
+
+content_splitter = TokenTextSplitter(chunk_size=1500, chunk_overlap=0)
+split_content = content_splitter.split_text(page_content)
+```
+
+We do a similar thing to get the last sentence or so from the previous loop, if there was one, so that ChatGPT can 
+stitch them together without grammar issues. 
+
+#### How much does it cost
+
+Well, we scraped about 10,000 pages before we got bored and started a different project. That took around 24 hours 
```

Comment:
Funny, but I think it sends the wrong message. I think it would be better to say "Until we hit our usage quota and needed to increase it"